Blog Post: Azure Front Door end-to-end TLS with AKS App Routing (#4825)
Azure/AKS: blog/_posts/2025-02-28-afd-aks-ingress-tls-approuting.md

On the backend, Nginx Ingress is an easy-to-use Kubernetes ingress controller. [AKS Application Routing](https://learn.microsoft.com/en-us/azure/aks/app-routing) provides a managed experience for deploying and managing Nginx Ingress in your AKS clusters. By pairing Azure Front Door with AKS App Routing, you get one of the easiest operational experiences for deploying and managing globally distributed web endpoints with content delivery optimization and security.

In this walkthrough we'll deploy an app with end-to-end TLS encryption, using Azure Front Door as the Internet Facing TLS endpoint and AKS App Routing Ingress as the internal in-cluster ingress controller. 

We'll use Azure Key Vault to store the TLS certificate, and will use App Routing Key Vault integration to get the certificate secret into the ingress controller. 

> *Note:* We'll be walking through step by step, showing the moving parts, via the Azure CLI. If you'd like to see a full implementation of a very similar deployment using Bicep, check out the link below:

> [End-to-end TLS with AKS, Azure Front Door, Azure Private Link Service, and NGINX Ingress Controller](https://learn.microsoft.com/en-us/samples/azure-samples/aks-front-door-end-to-end-tls/aks-front-door-end-to-end-tls/)

Author: Added a pre-req section. I did have the note on certbot installation further down in the doc, but I did add the need to be able to deploy a virtual machine with a public IP and access to port 80 as an optional requirement.

Let's get to it....

## Network Setup

First, we'll need to establish the network where our AKS cluster will be deployed. Nothing special in our network design. Just a Vnet with a single subnet for the cluster.

```bash
# Set starting environment variables
RG=E2ETLSLab
LOC=eastus2

# Create the Resource Group
az group create -g $RG -l $LOC

# Get the resource group id
RG_ID=$(az group show -g $RG -o tsv --query id)

# Set environment variables for the VNet creation
VNET_NAME=lablab-vnet
VNET_ADDRESS_SPACE=10.140.0.0/16
AKS_SUBNET_ADDRESS_SPACE=10.140.0.0/24

> # Create the Vnet along with the initial subet for AKS

Reviewer: spelling on subnet # Create the Vnet along with the initial subet for AKS -> # Create the Vnet along with the initial subnet for AKS

Reviewer: Everything else LGTM.

az network vnet create \
-g $RG \
-n $VNET_NAME \
--address-prefix $VNET_ADDRESS_SPACE \
--subnet-name aks \
--subnet-prefix $AKS_SUBNET_ADDRESS_SPACE 

# Get a subnet resource ID, which we'll need when we create the AKS cluster
VNET_SUBNET_ID=$(az network vnet subnet show -g $RG --vnet-name $VNET_NAME -n aks -o tsv --query id)
```

## Cluster Creation

Now, let's create the AKS cluster. This will be a very plain AKS cluster, for simplicity, however we will deploy to our above created subnet and will enable the following features:

- *AKS App Routing:* This will deploy managed Nginx Ingress

> - *App Routing Default Nginx Controller:* This allows us to ensure that the default deployment of App Routing uses a private IP

Reviewer 2: ```suggestion
- *App Routing Default Nginx Controller:* Type set to Internal which allows us to ensure that the default deployment of App Routing uses a private IP
```

- *Key Vault CSI Driver:* The Key Vault CSI driver will be used by App Routing to access our certificate in Azure Key Vault.

```bash
# NOTE: Make sure you give your cluster a unique name
CLUSTER_NAME=e2etlslab

# Cluster Creation Command
az aks create \
-g $RG \
-n $CLUSTER_NAME \
--node-count 2 \
--enable-app-routing \
--app-routing-default-nginx-controller Internal \
--enable-addons azure-keyvault-secrets-provider \

> --vnet-subnet-id $VNET_SUBNET_ID

Reviewer 2: Adding this flag in case folks don't already have SSH keys to use

```suggestion
--vnet-subnet-id $VNET_SUBNET_ID \
--generate-ssh-keys
```

```bash
# Create a key vault name
KEY_VAULT_NAME=e2elab$RANDOM

# Create the key value
az keyvault create --name $KEY_VAULT_NAME --resource-group $RG --location $LOC --enable-rbac-authorization false
```

For Azure Front door I'll need a public certificate for my hostname. If I were doing this for a company, I would reach out to the team that owns public certificates, but in my case, I'll just use [Certbot](https://certbot.eff.org/) to generate a free certificate from [LetsEncrypt](https://letsencrypt.org/).

I'm not going to get into all the specifics of using CertBot with LetsEncrypt, but the basics are as follows. The domain I'll be using is my 'crashoverride.nyc' domain.

1. Get an internet reachable host capable of running a webserver on ports 80 and 443
2. Create an A-record for your target domain to the public IP of the server. This is required for hostname validation used by Certbot
3. Run the certbot command as a privileged user on that web server host mentioned in #1 above

Here's a sample of the command I used to create the certificate:

```bash
sudo certbot certonly --key-type rsa --standalone -d e2elab.crashoverride.nyc
```

> Certbot creates several files, all with the PEM file extension. This is misleading, as fullchain.pem is the 'crt' file and the privkey.pem is the 'key' file. To store these in Azure Key Vault as certs we'll need to package these files up in a PFX format.

Reviewer 2: Might want to drop a link somewhere here to help users install certbot if they don't already have it. https://certbot.eff.org/instructions

> Certbot creates several files, all with the PEM file extension. This is misleading, as fullchain.pem is the 'crt' file and the privkey.pem is the 'key' file. To store these in Azure Key Vault as certs we'll need to package these files up in a PFX format.

Reviewer 2: Might want to drop a link somewhere here to help users install certbot if they don't already have it. https://certbot.eff.org/instructions

Author: Added a link to the install steps and parameterized the hostname in a few more places. Thanks!

```bash
# Create the managed identity
az identity create --name afd-identity --resource-group $RG --location $LOC

# Get identity client ID
export USER_ASSIGNED_CLIENT_ID=$(az identity show --resource-group $RG --name afd-identity --query 'clientId' -o tsv)
export USER_ASSIGNED_RESOURCE_ID=$(az identity show --resource-group $RG --name afd-identity --query 'id' -o tsv)

# Set the key vault access rights for the afd managed identity
az keyvault set-policy --name $KEY_VAULT_NAME -g $RG --secret-permissions get --spn "${USER_ASSIGNED_CLIENT_ID}"

# Create the Azure Front Door
az afd profile create \
--profile-name e2elab \
--resource-group $RG \
--identity-type UserAssigned \
--user-assigned-identities $USER_ASSIGNED_RESOURCE_ID \
--sku Premium_AzureFrontDoor 
```

We'll do the rest in our Azure Front Door instance in the Azure Portal, so open a browser to [https://portal.azure.com](https://portal.azure.com) and then navigate to your resource group and then your Azure Front Door Instance.

### Link the certificate to the AFD.

> To use our certificate with Azure Front Door, we need to attach the certificate in Azure Key Vault to an Front Door Secret. We do this in the 'Secrets' pane under 'Security'.

Reviewer 2: ```suggestion
To use our certificate with Azure Front Door, we need to attach the certificate in Azure Key Vault to a Front Door Secret. We do this in the 'Secrets' pane under 'Security'.
```
